### PR TITLE
gate: add gate holders debugging

### PR DESCRIPTION
**Pull request by bhalevy**
gate::holder keeps a naked pointer to the gate it holds. Its life time is independent of the gate and there is no mechanism to update this pointer if the gate is moved, therefore if the gate is moved or destroyed while there are outstanding holders pointing to it there is a risk for use-after-move or use-after-free.

Count the number of gate holders in debug mode
and assert it's zero when the gate is moved or destroyed.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

### CODE REVIEW

**Comment by bhalevy** [comment]
Cc @tchaikov

**Comment by bhalevy** [comment]
CI failure seems completely unrelated...

File: include/seastar/core/gate.hh [hunk 1]
```diff
@@ -25,6 +25,10 @@
 #include <seastar/util/std-compat.hh>
 #include <exception>
 
+#ifdef SEASTAR_DEBUG
```
Comment (tchaikov) [inline include/seastar/core/gate.hh:28]: why not just use `SEASTAR_DEBUG` ? do we expect user to use `GATE_HOLDER_DEBUG` directly?

File: include/seastar/core/gate.hh [hunk 2]
```diff
@@ -47,21 +51,36 @@ public:
 class gate {
     size_t _count = 0;
     std::optional<promise<>> _stopped;
+
+#ifdef GATE_HOLDER_DEBUG
+    size_t _holders = 0;
+
+    void assert_not_held(const gate& g) const noexcept {
```
Comment (tchaikov) [inline include/seastar/core/gate.hh:58]: probably could just take advantage of the fact that this is a member function? like
```c++
void assert_not_held() const noexcept {
    assert(!_holders && "gate moved with outstanding holders");
}
```

File: include/seastar/core/gate.hh [hunk 3]
```diff
@@ -47,21 +51,36 @@ public:
...
+        assert(!g._holders && "gate moved with outstanding holders");
+    }
+#else   // GATE_HOLDER_DEBUG
+    void assert_not_held(const gate& g) const noexcept {}
```
Comment (tchaikov) [inline include/seastar/core/gate.hh:62]: so we don't need to worry about a picky compiler which might warn on a defined-but-not used variable.

File: include/seastar/core/gate.hh (hunk 1 above)
Comment (bhalevy) [reply include/seastar/core/gate.hh:28]: I'd like to reserve the opportunity to control it in some other way (configure/cmake option, !NDEBUG, etc.)

File: include/seastar/core/gate.hh (hunk 2 above)
Comment (bhalevy) [reply include/seastar/core/gate.hh:58]: good point. will do

**Comment by bhalevy** [comment]
In v2 ([6e05287](https://github.com/scylladb/seastar/commit/6e05287c7e0454c09958be91f65439d10b19d525)):
- used assert_not_held as member function

**Review by tchaikov — APPROVED** [review]

File: include/seastar/core/gate.hh (hunk 1 above)
Comment (tchaikov) [reply include/seastar/core/gate.hh:28]: ack.

**Comment by bhalevy** [comment]
@avikivity what's this failure in spawn_test with g++-12?
https://app.circleci.com/pipelines/github/bhalevy/seastar/63/workflows/f38464f8-c939-4194-b4f2-95816568f66c/jobs/584

**Comment by tchaikov** [comment]
@bhalevy Benny, this is a known issue. see #1320. i am on it.

File: include/seastar/core/gate.hh [hunk 4]
```diff
@@ -25,6 +25,10 @@
 #include <seastar/util/std-compat.hh>
 #include <exception>
 
+#ifdef SEASTAR_DEBUG
+#define GATE_HOLDER_DEBUG
```
Comment (avikivity) [inline include/seastar/core/gate.hh:29]: SEASTAR_ prefix to not pollute the global namespace

Comment (bhalevy) [reply include/seastar/core/gate.hh]: fixed in v3

**Comment by bhalevy** [comment]
In v3 ([5a0b552](https://github.com/scylladb/seastar/commit/5a0b55251c1c1fb8d5c3efd24f6d6e5ec16ce771)):
- s/GATE_HOLDER_DEBUG/SEASTAR_GATE_HOLDER_DEBUG/

File: tests/unit/abort_source_test.cc [hunk 5]
```diff
@@ -161,3 +161,14 @@ SEASTAR_THREAD_TEST_CASE(test_sleep_abortable_with_exception) {
     }
     BOOST_REQUIRE(caught_exception);
 }
+
+SEASTAR_TEST_CASE(test_destroy_subscribed_abort_source) {
```
Comment (xemul) [inline tests/unit/abort_source_test.cc:165]: Why is abort-source test here? The abort_source doesn't have any gates inside.

Comment (bhalevy) [reply tests/unit/abort_source_test.cc:165]: (Relevant for https://github.com/scylladb/scylladb/issues/12183)

**Comment by xemul** [comment]
@tchaikov , from this CI run https://app.circleci.com/pipelines/github/bhalevy/seastar/74/workflows/692df47b-9477-4a74-bf58-d66497d3fe36/jobs/649

```
72/74 Test #72: Seastar.unit.spawn ............................***Failed    1.08 sec
[0/1] cd /home/circleci/project/build/debug/tests/unit && /home/circleci/project/build/debug/tests/unit/spawn_test -- -c 2
Running 5 test cases...
WARNING: debug mode. Not for benchmarking or production
INFO  2022-12-06 10:42:14,971 seastar - Reactor backend: linux-aio
WARN  2022-12-06 10:42:14,982 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
WARN  2022-12-06 10:42:14,982 [shard 1] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
INFO  2022-12-06 10:42:14,996 [shard 0] seastar - Created fair group io-queue-0, capacity rate 2147483:2147483, limit 12582912, rate 16777216 (factor 1), threshold 2000
INFO  2022-12-06 10:42:14,998 [shard 0] seastar - IO queue uses 0.75ms latency goal for device 0
INFO  2022-12-06 10:42:15,000 [shard 0] seastar - Created io group dev(0), length limit 4194304:4194304, rate 2147483647:2147483647
INFO  2022-12-06 10:42:15,000 [shard 0] seastar - Created io queue dev(0) capacities: 512:2000:2000 1024:3000:3000 2048:5000:5000 4096:9000:9000 8192:17000:17000 16384:33000:33000 32768:65000:65000 65536:129000:129000 131072:257000:257000
random-seed=3376081796
unknown location(0): fatal error: in "test_spawn_input": std::system_error: Broken pipe
/home/circleci/project/src/testing/seastar_test.cc(43): last checkpoint
```

it doesn't seem related to this change

**Comment by tchaikov** [comment]
> @tchaikov , from this CI run https://app.circleci.com/pipelines/github/bhalevy/seastar/74/workflows/692df47b-9477-4a74-bf58-d66497d3fe36/jobs/649
> 
> ```
> 72/74 Test #72: Seastar.unit.spawn ............................***Failed    1.08 sec
> [0/1] cd /home/circleci/project/build/debug/tests/unit && /home/circleci/project/build/debug/tests/unit/spawn_test -- -c 2
> Running 5 test cases...
> WARNING: debug mode. Not for benchmarking or production
> INFO  2022-12-06 10:42:14,971 seastar - Reactor backend: linux-aio
> WARN  2022-12-06 10:42:14,982 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
> WARN  2022-12-06 10:42:14,982 [shard 1] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
> INFO  2022-12-06 10:42:14,996 [shard 0] seastar - Created fair group io-queue-0, capacity rate 2147483:2147483, limit 12582912, rate 16777216 (factor 1), threshold 2000
> INFO  2022-12-06 10:42:14,998 [shard 0] seastar - IO queue uses 0.75ms latency goal for device 0
> INFO  2022-12-06 10:42:15,000 [shard 0] seastar - Created io group dev(0), length limit 4194304:4194304, rate 2147483647:2147483647
> INFO  2022-12-06 10:42:15,000 [shard 0] seastar - Created io queue dev(0) capacities: 512:2000:2000 1024:3000:3000 2048:5000:5000 4096:9000:9000 8192:17000:17000 16384:33000:33000 32768:65000:65000 65536:129000:129000 131072:257000:257000
> random-seed=3376081796
> unknown location(0): fatal error: in "test_spawn_input": std::system_error: Broken pipe
> /home/circleci/project/src/testing/seastar_test.cc(43): last checkpoint
> ```
> 
> it doesn't seem related to this change

@xemul it's a known issue. and is tracked by #1320.

**Comment by bhalevy** [comment]
ping review

File: include/seastar/core/gate.hh [hunk 6]
```diff
@@ -47,21 +51,36 @@ public:
...
             assert(!_count && "gate reassigned with outstanding requests");
+            x.assert_not_held();
             _count = std::exchange(x._count, 0);
             _stopped = std::exchange(x._stopped, std::nullopt);
         }
         return *this;
     }
     ~gate() {
         assert(!_count && "gate destroyed with outstanding requests");
+        assert_not_held();
```
Comment (xemul) [inline include/seastar/core/gate.hh:83]: nit: this would print "gate moved ..." while technically the gate is not moved here

Comment (bhalevy) [reply include/seastar/core/gate.hh]: Fixed in [dc5afad](https://github.com/scylladb/seastar/commit/dc5afadbd5deec5c069029016fd8c399551b45f7)

**Comment by bhalevy** [comment]
In [dc5afad](https://github.com/scylladb/seastar/commit/dc5afadbd5deec5c069029016fd8c399551b45f7):
- rebased
- split assert_not_held to move and destroy cases

**Review by xemul — APPROVED** [review]

**Comment by bhalevy** [comment]
@scylladb/seastar-maint please consider merging

**Review by nyh — APPROVED** [review]

**Pull request closed without merging**